### ISSUE TITLE
Safari 6 - No such table: dbversions

### ISSUE THREAD

**Issue by napcs** [axemclion/IndexedDBShim#58]
In Safari 6 and on Mobile Safari iOS 6, when I first load my page, I get this error:

```
could not prepare statement (1 no such table: dbVersions)
```

If I refresh the page, everything works just fine.  My code is ridiculously simple:

```
var db = null;

var connectToDB = function(){
  var version = 1;  
  var request = indexedDB.open("awesomenotes", version);

  request.onupgradeneeded = function(event) {
    db = event.target.result;
    db.createObjectStore("notes", { keyPath: "id", autoIncrement: true });
  };

  request.onsuccess = function(event) {
    db = event.target.result;    
  };

  request.onerror = function(event){
    alert(event.debug[1].message);
  }
};

connectToDB();
```

I am loading the shim at the bottom of my page, before the above code.

If I move the shim into the `<head>` section, it works fine.

Is there a reason this has to be loaded in the head? Or is there something else going on?

**Comment by jtr3icio** [comment]
After looking at the source of the shim I think that the problem lies in the asynchronous way the database works. The first database call always goes to the system database, looking up the versions. Because this is an asynchronous operation it is added to the event loop.

If this fails an error handler will be executed, creating a new table in the database. Again this is asynchronous and will be done in another tick of the event loop. Because the code (except the asynchronous functions) has finished executing the browsers resumes parsing your code.

Here comes the problem: If there is nothing between your code and the shim the browser will just continue with your code. Now your code is in the same tick as the shim. Now you start another asynchronous operation, which will again be added to the eventloop.

But because this is still the first run it hasn't run the check for the versions-table (which the shim checks for) and if this check fails (after your code ran) it will add the error handler to the event loop, AFTER your attempt to open a database, which then will fail, because the shim isn't fully initialized.

If you put some HTML (or something else) between the shim and your script the check for the table will execute at sometime while parsing and rendering the html (the parser will stop, but it's to short to notice) and then continue parsing. This forces the browser to execute your code in another tick.

**Comment by gabrielmaldi** [comment]
I can confirm this. However, I couldn't make it work by adding some HTML between the shim and my script, no matter what I tried. I ended up having to wrap db.js inside a function which I call after a setTimeout of 1 second (if I use less than this interval, it fails).

I'd like to hear of a cleaner solution, or at least one that doesn't involve delaying the startup process of my app for 1 second.

Thanks

**Comment by jtr3icio** [comment]
The best solution would be a jQuery.ready like function, which takes a callback and executes it when the database is ready or directly if the database already exists. I'm busy right now, maybe someone could add such a function to the shim. I don't think that this would be that much work.

**Comment by axemclion** [comment]
I tried running this page on Safari 6 - http://nparashuram.com/IndexedDBShim/test/index.html and the tests seem to run fine.

**Comment by gabrielmaldi** [comment]
Totally agree with jtr3icio: we need a callback invoked when the shim is ready.

**Comment by DickvdBrink** [comment]
Fixed with #148
